### ISSUE TITLE
Implement support for multi-column content

### ISSUE THREAD

**Issue by Daniel-Miller** [InSite/Lotus#20]
I have moved the Markdown authoring application into the Lotus repo, and it is now renamed Lotus.Web.Author.  As a next step in the development of this prototype, there are some enhancements we should implement.

We have some potential customers who want to use this tool to author content that renders in multiple columns. The Markdown syntax for tables will not work -- mainly because it is too confusing, but also because our users want more than single-line rows in each column.

My idea is to introduce two "control strings": {Section-Break} and {Column-Break}

We can run the user's content through a preprocessing function before the markdown is transformed into HTML, and the preprocessing function can perform these steps:

1. Replace the first occurrence of a {Section-Break} with elements to open a table, tr, and td.

2. Replace the last occurrence of a {Section-Break} with elements to close the table, tr, and td. (If the first {Section-Break} is the only only one, then append closing elements at the end of the content.)

3. Replace {Column-Break} with elements to close the td, then open a new td.

4. When a {Section-Break} appears between the first and last {Section-Break}, replace it with elements to close the td and tr, then open a new tr and td.

Do you think this will work? Here is a mockup of the UI:

![mock](https://cloud.githubusercontent.com/assets/5367871/15683977/442f8418-2722-11e6-9eff-4fd354bf4652.png)

**Comment by Oleg-Terzi** [comment]
> Do you think this will work?

I think that this will work but I want to amplify a little the rules defined by you:

1. The first occurence of a {Section-Break} will be replaced with HTML table only if it has closing {Section-Break} and has {Column-Break} between these block; otherwise it will be ignored. For example, these {Section-Break} control strings will be ignored (removed):
    ![img_1](https://cloud.githubusercontent.com/assets/5382331/15870014/739a4226-2cf7-11e6-8b17-5bb3d7500e13.png)
    -
![img_2](https://cloud.githubusercontent.com/assets/5382331/15870021/7d287e8e-2cf7-11e6-8485-75cc166d3a8d.png)

2. The HTML table will be closed in case when the current {Section-Break} is the last or when between the current {Section-Break} and the next {Section-Break} are any {Column-Break}. For example:
![img_3](https://cloud.githubusercontent.com/assets/5382331/15870054/9e9f8300-2cf7-11e6-8746-89d5c61bde5e.png)

How do you think, is it OK?

**Comment by Oleg-Terzi** [comment]
The task is done and I have few notes about it:

1. I implemented here **post**-processing rather than **pre**-processing, because the Markdown processor doesn't work very well with HTML.
2. I propose to use here a DIV-based rendering (like grid system that is implemented in Bootstrap) rather than a TABLE-based rendering -- it seems to me that it is more suited to our purposes. Please let me know if we need to use TABLE-based rendering here.

And here is an example of what we get in output for now:
![img_1](https://cloud.githubusercontent.com/assets/5382331/15902270/ccbac282-2daf-11e6-9dff-e5fbe4610093.png)

**Comment by Daniel-Miller** [comment]
A div-based layout is definitely better for web page content, but worse for email message content. Perhaps we could make this conditional, based on the content type selected by the user?

**Comment by Daniel-Miller** [comment]
Your implementation looks great so far. I think we will need table-based rendering for email message content. If it makes the programming simpler, we can assume an email message contains one column only (i.e. no column breaks), or two columns (e.g. English and French), and never more than two columns.

**Comment by Oleg-Terzi** [comment]
Done.

> If it makes the programming simpler, we can assume an email message contains one column only (i.e. no column breaks), or two columns (e.g. English and French), and never more than two columns.

Here are no problems with many columns. The only difficulty here is to determine the end of table because we do not have the closing "tag". For example, if an user wants to render page like this:

{TABLE.1:START}
&nbsp;&nbsp;&nbsp;&nbsp;{ROW.1}
{TABLE.1:END}

{TEXT}

{TABLE.2:START}
&nbsp;&nbsp;&nbsp;&nbsp;{ROW.1}
{TABLE.2:END}

We can't determine where is end of first and start of second table and this structure will be rendered as one table:

{TABLE.1:START}
&nbsp;&nbsp;&nbsp;&nbsp;{ROW.1}
&nbsp;&nbsp;&nbsp;&nbsp;{ROW.2:START}
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;{TEXT}
&nbsp;&nbsp;&nbsp;&nbsp;{ROW.2:END}
&nbsp;&nbsp;&nbsp;&nbsp;{ROW.3}
{TABLE.1:END}
    
In other words, the current syntax allows us to create only one 'Section-Break' table per document.
    
Considering this I supposed that the each '{Section-Break} ... {Section-Break}' structure will be rendered as separate HTML table and we can't create a multi-row table at this moment.

**Comment by Daniel-Miller** [comment]
We can assume that our users will not need multiple tables in an Email Message, but some users might need a table with multiple rows (2 columns only). 

For example:

{Row.1}
   English Content
   {Column-Break}
   French Content
{Row.2}
   English Content
   {Column-Break}
   French Content
{Row.3}
   English Content
   {Column-Break}
   French Content

They do this because they want to ensure that content sections are horizontally aligned, even when the size of the content for French is larger than the size of the content for English.  (Generally, when an English sentence is translated to French, the French equivalent contains a longer string of text.)

Can you think of a simple and elegant solution to this? 

I don't want to ask our users to insert starting and ending tags for tables and rows, because I think that's too complicated for them.  Our users always find this confusing.

This is the reason I am hoping we can do something like this:

   English Content
   {Column-Break}
   French Content
{Section-Break}
   English Content
   {Column-Break}
   French Content
{Section-Break}
   English Content
   {Column-Break}
   French Content

Maybe we prompt the user for the number of columns, and our code always assumes one table?

**Comment by Oleg-Terzi** [comment]
Done.

> Can you think of a simple and elegant solution to this?

I decided to implement here the next solution -- suppose a user has the next document:

![img_1](https://cloud.githubusercontent.com/assets/5382331/15987004/27095af4-3022-11e6-92c5-207e100c3b97.png)

As you can see the markdown content is rendered there as a table with three rows. Suppose a user wants to place the first and the last rows to separate tables and render middle row as a simple text. To do this a user should add second empty new line after the last {Section-Break} of first table:

![img_2](https://cloud.githubusercontent.com/assets/5382331/15987005/2c888bee-3022-11e6-9d27-b7dbd4f5f939.png)

For me it looks like an intuitive way to mark the end of the table, but I'm not sure that it will be intuitive for users. What do you think about this solution?

**Comment by Daniel-Miller** [comment]
This looks really good, and it's simple for our users. I'll ask them do some review and testing, and give us their feedback.

**Comment by Daniel-Miller** [comment]
I have added some new CSS rules to Style 1, and the preview isn't rendering the way it should.

![preview](https://cloud.githubusercontent.com/assets/5367871/16496700/7a12bc44-3eb0-11e6-88b3-19707124dc9c.png)

I think our approach to tables is probably making our solution too complex, and maybe there is a better and simpler solution?

The only context in which we need to organize content into a table with multiple rows and columns is an email message that is sent to recipients where the sender is required to communicate in two official languages (English and French in Canada).

In other words, the content will have never have more than two columns, and the content will never have a two column row followed by a one column row.

This means we could delete the {Column-Break} placeholder, and instead introduce a {Language-Break} placeholder.  This placeholder will not occur or it will occur once; it should never occur more than once.

What do you think of an approach like this?

![example](https://cloud.githubusercontent.com/assets/5367871/16496705/83d271ac-3eb0-11e6-9079-170c3f81ec99.png)

**Comment by Oleg-Terzi** [comment]
Done. I implemented the syntax for Section-Break in the way you discribed it. Just in case, I left the old Document Type options (they marked by 'obsolete' label) and if new syntax is satisfies to our needs then we could delete the old syntax completely.

> I have added some new CSS rules to Style 1, and the preview isn't rendering the way it should.

The description of Issue #24 says that we should apply a Section-Break class (":identifier") to the rendered table:

```
This will include a CSS class attribute in the replacement HTML. For example:

<table class="signature"><tr><td>
```

If we divide the Section-Break on two tables then it would work as you expected. For example:

![img_1](https://cloud.githubusercontent.com/assets/5382331/16529027/26647582-3fc9-11e6-96f0-b4142b0d570b.png)

For new syntax it will not work because we can create only one table in a document. So if I understand it correct, we need to change this feature so that a Section-Break class will be applied to a row instead of a table. I made relevant changes and your example is rendered in the next way:

1. For old syntax:
    ![img_2](https://cloud.githubusercontent.com/assets/5382331/16529032/2c0f40a2-3fc9-11e6-9efc-1e64b86a9c8d.png)

2. For new syntax:
    ![img_3](https://cloud.githubusercontent.com/assets/5382331/16529038/354d6748-3fc9-11e6-876e-7200f8494f8a.png)

**Comment by Daniel-Miller** [comment]
This looks good; I'll get our users to review and test. The only change we need now is to set the width for each of the two columns on an Email Message to 50%. This will ensure they have the same width and the same horizontal alignment.  

This is especially important to some of our bilingual customers, who are required to ensure that neither official language *appears* to have preferential treatment in the layout of a page or in the layout of an email message.

**Comment by Oleg-Terzi** [comment]
Done.
